### PR TITLE
[SYClomatic][CodePin]Refine CodePin by switching from using schema to directly recording data with functions

### PR DESCRIPTION
**Pull request by wenhuiNi**

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp [hunk 1]
```diff
@@ -0,0 +1,226 @@
+//==---- dispatcher.hpp -------------------------------*- C++
+//-*----------------==//
```
Comment (wenhuiNi) [inline clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: format

File: clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  if (log.back() == ',')
+    log.pop_back(); // Pop last ',' character
+  dump_data(new_api_name, log);
+}
+
+} // namespace detail
+} // namespace experimental
+} // namespace dpct
+
+#endif
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: no new line

File: clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+#ifdef __NVCC__
+inline void process_var(std::string log, cudaStream_t stream) { ; }
+#else
+inline void process_var(std::string log, dpct::queue_ptr stream) { ; }
+#endif
+
+template <class T, class... Args>
+#ifdef __NVCC__
+void process_var(std::string log, cudaStream_t stream,
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: may add a macro define of stream:
```
#ifdef __NVCC__
typedef StreamType cudaStream_t;
#else
typedef StreamType dpct::queue_ptr;
#endif
```

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 4]
```diff
@@ -26,18 +25,17 @@ inline void synchronize(cudaStream_t stream) { cudaStreamSynchronize(stream); }
...
 /// \param api_name The UID of the function call.
 /// \param stream The CUDA stream to synchronize the command execution.
-/// \param args The schema string and variable value pair list.
+/// \param args The var name string and variable value pair list.
 template <class... Args>
 void gen_prolog_API_CP(const std::string &api_name, cudaStream_t stream,
-                       Args... args) {
-  synchronize(stream);
-  dpct::experimental::detail::gen_log_API_CP(api_name, args...);
+                       Args... args) synchronize(stream);
```
Comment (ShengchenJ) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: Missed left curly brackets
```suggestion
                       Args... args) {
                       synchronize(stream);
```

File: clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+namespace experimental {
+namespace detail {
+
+inline static std::map<std::string, int> api_index;
+inline static std::vector<std::string> dump_json;
+inline static std::string dump_file = "dump_log.json";
+
+class Logger {
+public:
+  Logger(const std::string &dumpFile)
```
Comment (ShengchenJ) [inline clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: Let's use the underscore name format to align with the old header file.

File: clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+inline uint32_t get_pointer_size_in_bytes_from_map(void *ptr) {
+  const std::map<void *, uint32_t> &ptr_size_map = get_ptr_size_map();
+  const auto &it = ptr_size_map.find(ptr);
+  return (it != ptr_size_map.end()) ? it->second : 0;
+}
+
+#ifdef __NVCC__
+void dump(std::string &log, const int3 &value) {
+  log = log + '\t' + std::to_string(value.x);
```
Comment (ShengchenJ) [inline clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: How about use the "
  std::ostringstream oss;
  oss << '\t' << value.x
      << '\t' << value.y
      << '\t' << value.z;
  
  log += oss.str();"

File: clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+  attributes.init(p);
+  if (attributes.get_device_pointer() != nullptr)
+    return true;
+  return false;
+#endif
+}
+
+template <class T>
+#ifdef __NVCC__
+void dispatch_codePin(std::string &log, T a, cudaStream_t stream) {
```
Comment (ShengchenJ) [inline clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]:
```suggestion
void dispatch_codepin(std::string &log, T a, cudaStream_t stream) {
```

File: clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+#ifdef __NVCC__
+void dump(std::string &log, const int3 &value) {
+  log = log + '\t' + std::to_string(value.x);
+  log = log + '\t' + std::to_string(value.y);
+  log = log + '\t' + std::to_string(value.z);
+}
+
+void dump(std::string &log, const float3 &value) {
```
Comment (AndyCHHuang) [inline clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: use stream?

Comment (AndyCHHuang) [reply clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: prefix...
dump(value.x);
dump(value.y);
...
postfix..

File: clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+void dump(std::string &log, const sycl::float3 &value) {
+  log = log + '\t' + std::to_string(value.x());
+  log = log + '\t' + std::to_string(value.y());
+  log = log + '\t' + std::to_string(value.z());
+}
+#endif
+
+void dump(std::string &log, const char *value) {
```
Comment (AndyCHHuang) [inline clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: dump(string_view)

File: clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+#endif
+}
+
+template <class T>
+#ifdef __NVCC__
+void dispatch_codePin(std::string &log, T a, cudaStream_t stream) {
+#else
+void dispatch_codePin(std::string &log, T a, dpct::queue_ptr stream) {
+#endif
+  if constexpr (std::is_pointer_v<T>) {
```
Comment (AndyCHHuang) [inline clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: can we use overloading instead of if constexpr (std::is_pointer_v<T>) {

Comment (AndyCHHuang) [reply clang/runtime/dpct-rt/include/dpct/codepin/detail/dispatcher.hpp]: maybe we dont need dispatch
just overloading dump()

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 11]
```diff
@@ -8,50 +8,164 @@
...
+};
+
+inline static Logger log(dump_file);
+
+inline std::map<void *, uint32_t> &get_ptr_size_map() {
+  static std::map<void *, uint32_t> ptr_size_map;
+  return ptr_size_map;
+}
+
+inline uint32_t get_pointer_size_in_bytes_from_map(void *ptr) {
```
Comment (ShengchenJ) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: Please align  the pointer or ptr with "get_ptr_size_map" name

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 12]
```diff
@@ -8,50 +8,164 @@
...
-namespace experimental {
+}
 
+template <class T>
+class TT<T, typename std::enable_if<std::is_pointer<T>::value>::type> {
+public:
+  static void dump(std::ostream &ss, T value,
+                   dpct::experimental::StreamType stream) {
+    int size = get_pointer_size_in_bytes_from_map(value);
+    size = size == 0 ? 1 : size / sizeof(*value);
```
Comment (ShengchenJ) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp:104]: The size in the get_pointer_size_in_bytes_from_map return and the count of the T is different in this context, I prefer to use a new variable here.

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 13]
```diff
@@ -8,50 +8,164 @@
...
+    using PointedType =
+        std::remove_reference_t<std::remove_cv_t<std::remove_pointer_t<T>>>;
+    if (is_dev_ptr(value)) {
+      PointedType *hData = new PointedType[size];
```
Comment (ShengchenJ) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: Align the format, use the h_data.

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 14]
```diff
@@ -8,50 +8,164 @@
...
+      api_name + ":" + std::to_string(api_index[api_name]);
+  log.get_outputstream() << "{\"ID\":"
+                         << "\"" << new_api_name << "\",\"CheckPoint\":{";
+  process_var(log.get_outputstream(), stream, args...);
+  log.get_outputstream() << "}}";
 }
+} // namespace detail
+
+#ifdef __NVCC__
+inline void synchronize(cudaStream_t stream) { cudaStreamSynchronize(stream); }
```
Comment (ShengchenJ) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp:185]: Replace with the StreamType

File: clang/runtime/dpct-rt/include/dpct/codepin/serialization/serial_basic.hpp [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+                   dpct::experimental::StreamType stream) {
+    ss << value << ",\t";
+  }
+};
+
+} // namespace detail
+} // namespace experimental
+} // namespace dpct
+
+#endif
```
Comment (ShengchenJ) [inline clang/runtime/dpct-rt/include/dpct/codepin/serialization/serial_basic.hpp]: Add a empty line at the end of file.

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by intwanghao — APPROVED** [review]
LGTM

**Review by ShengchenJ — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 16]
```diff
@@ -1,83 +1,212 @@
...
+#include <unordered_set>
+#include <utility>
+#include <vector>
+namespace dpct {
+namespace experimental {
+
+namespace detail {
+inline static std::unordered_set<void *> ptr_unique;
+inline static std::map<std::string, int> api_index;
+inline static std::string dump_file = "dump_log.json";
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: Please add comment for the dump_file, including what kinds of content it has. 
Also consider using a dynamic name, likes: 
option 1: ENV variable
option 2: Get the APP binary name, and use it as the part name of the log file.
option 3: Include a rand number based on time as part name of the log file.
option 4:  insert the basic filename defined in CPP code for both CUDA/SYCL code base.
 may be option3 and option4,  option2 + option4  worth a try.

Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]:
```suggestion
inline static std::string data_file = "app_runtime_data_record.json"; // default name
```

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 17]
```diff
@@ -1,83 +1,212 @@
...
+  ~Logger() {
+    this->remove_lastchar_stream();
+    ss << "]";
+    opf << ss.str();
+    opf.close();
+  }
+
+  std::stringstream &get_outputstream() { return this->ss; }
+
+  void remove_lastchar_stream() {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp:49]: add a comment to explain the purpose.

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 18]
```diff
@@ -1,83 +1,212 @@
...
+    ss << "";
+  }
+
+private:
+  std::string dst_output;
+  std::ofstream opf;
+  std::stringstream ss;
+};
+
+inline static Logger log(dump_file);
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: Please put all the static member into same block.

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 19]
```diff
@@ -1,83 +1,212 @@
...
+
+inline std::map<void *, uint32_t> &get_ptr_size_map() {
+  static std::map<void *, uint32_t> ptr_size_map;
+  return ptr_size_map;
+}
+
+inline uint32_t get_ptr_size_in_bytes_from_map(void *ptr) {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]:
```suggestion
inline uint32_t get_ptr_size_in_bytes(void *ptr) {
```

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 20]
```diff
@@ -1,83 +1,212 @@
...
-void gen_prolog_API_CP(const std::string &api_name, cudaStream_t stream,
-                       Args... args) {
-  synchronize(stream);
-  dpct::experimental::detail::gen_log_API_CP(api_name, args...);
+    ss << "]}";
+  }
+};
+
+template <class T>
+class TT<T, typename std::enable_if<std::is_array<T>::value>::type> {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: How about change TT to DataSer / (Data serialize for any type of Data),  and add comment for the util class.

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp [hunk 21]
```diff
@@ -1,83 +1,212 @@
...
+  }
+};
+
+inline void process_var(std::ostream &ss,
+                        dpct::experimental::StreamType stream) {
+  ;
+}
+
+template <class T, class... Args>
+void process_var(std::ostream &ss, dpct::experimental::StreamType stream,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]:
```suggestion
void serialize_var(std::ostream &ss, dpct::experimental::StreamType stream,
```

File: clang/runtime/dpct-rt/include/dpct/codepin/serialization/serial_basic.hpp [hunk 22]
```diff
@@ -0,0 +1,155 @@
+//==---- serial_basic.hpp ----------------------------*-C++-*-------------==//
+//
+// Copyright (C) Intel Corporation
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===--------------------------------------------------------------------===//
+#ifndef __DPCT_CODEPIN_SERIAL_BASIC_HPP__
+#define __DPCT_CODEPIN_SERIAL_BASIC_HPP__
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/serialization/serial_basic.hpp]:
```suggestion
#define __DPCT_CODEPIN_SER_BASIC_HPP__
```

File: clang/runtime/dpct-rt/include/dpct/codepin/serialization/serial_basic.hpp [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+namespace detail {
+
+template <class T, class T2 = void> class TT {
+public:
+  static void dump(std::ostream &ss, T value,
+                   dpct::experimental::StreamType stream) {
+    ss << "{\"Type\":\"" << typeid(T).name() << "\",\"Data\":[";
+    ss << "Unable to find the corresponding serialization function for the "
+          "class. Please report this issue to SYCLomatic/DPCT, or manually add "
+          "the serialization function.";
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/serialization/serial_basic.hpp]: try to define the error code for L38,  1. as the logfile is expected to be quite big, not readable for most fields.  2.make compare script easy.  

At same time,  emit a message to tell the end user we have problem here when dump.

File: clang/runtime/dpct-rt/include/dpct/codepin/serialization/serial_basic.hpp [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+#endif
+
+namespace detail {
+
+template <class T, class T2 = void> class TT {
+public:
+  static void dump(std::ostream &ss, T value,
+                   dpct::experimental::StreamType stream) {
+    ss << "{\"Type\":\"" << typeid(T).name() << "\",\"Data\":[";
+    ss << "Unable to find the corresponding serialization function for the "
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/serialization/serial_basic.hpp]:
```suggestion
    ss << "Dump data error:  Unable to find the corresponding serialization function for the class: ${class_name}.  Skipped. \n";
```

File: clang/runtime/dpct-rt/include/dpct/codepin/serialization/basic.hpp [hunk 25]
```diff
@@ -30,31 +30,47 @@ typedef dpct::queue_ptr StreamType;
...
 namespace detail {
 
-template <class T, class T2 = void> class TT {
+template <typename T> void demangle_name(std::ostream &ss) {
+  int s;
+  auto mangle_name = typeid(T).name();
+  auto demangle_name = abi::__cxa_demangle(mangle_name, NULL, NULL, &s);
+  if (s != 0) {
+    ss << "CODEPIN:0: Unable to demangle symbol " << mangle_name
+       << ". Please report this issue to SYCLomatic/DPCT.";
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/serialization/basic.hpp]: let's remove L39.

File: clang/runtime/dpct-rt/include/dpct/codepin/serialization/basic.hpp [hunk 26]
```diff
@@ -30,31 +30,47 @@ typedef dpct::queue_ptr StreamType;
 
 namespace detail {
 
-template <class T, class T2 = void> class TT {
+template <typename T> void demangle_name(std::ostream &ss) {
+  int s;
+  auto mangle_name = typeid(T).name();
+  auto demangle_name = abi::__cxa_demangle(mangle_name, NULL, NULL, &s);
+  if (s != 0) {
+    ss << "CODEPIN:0: Unable to demangle symbol " << mangle_name
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/serialization/basic.hpp]:
```suggestion
    ss << "CODEPIN:ERROR:0: Unable to demangle symbol " << mangle_name
```

File: clang/runtime/dpct-rt/include/dpct/codepin/serialization/basic.hpp [hunk 27]
```diff
@@ -30,31 +30,47 @@ typedef dpct::queue_ptr StreamType;
...
   static void dump(std::ostream &ss, T value,
                    dpct::experimental::StreamType stream) {
-    ss << "{\"Type\":\"" << typeid(T).name() << "\",\"Data\":[";
-    ss << "Unable to find the corresponding serialization function for the "
+    ss << "{\"Type\":\"";
+    demangle_name<T>(ss);
+    ss << "\",\"Data\":[";
+    ss << "CODEPIN:1: Unable to find the corresponding serialization function "
+          "for the "
           "class. Please report this issue to SYCLomatic/DPCT, or manually add "
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/serialization/basic.hpp]: remove "please ......" statment.

File: clang/runtime/dpct-rt/include/dpct/codepin/serialization/basic.hpp [hunk 28]
```diff
@@ -30,31 +30,47 @@ typedef dpct::queue_ptr StreamType;
...
+template <class T, class T2 = void> class DataSer {
 public:
   static void dump(std::ostream &ss, T value,
                    dpct::experimental::StreamType stream) {
-    ss << "{\"Type\":\"" << typeid(T).name() << "\",\"Data\":[";
-    ss << "Unable to find the corresponding serialization function for the "
+    ss << "{\"Type\":\"";
+    demangle_name<T>(ss);
+    ss << "\",\"Data\":[";
+    ss << "CODEPIN:1: Unable to find the corresponding serialization function "
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/codepin/serialization/basic.hpp]:
```suggestion
    ss << "CODEPIN:ERROR:1: Unable to find the corresponding serialization function "
```

File: clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp (hunk 16 above)
Comment (intwanghao) [reply clang/runtime/dpct-rt/include/dpct/codepin/codepin.hpp]: Will do in another pr

**Review by ShengchenJ — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**